modin/pandas/groupby.py

```diff
@@ -128,7 +129,46 @@ def sem(self, ddof=1):
         return self._default_to_pandas(lambda df: df.sem(ddof=ddof))
 
     def mean(self, *args, **kwargs):
-        return self._apply_agg_function(lambda df: df.mean(*args, **kwargs))
+        fallback = False
+        converted_columns = {}
+        numeric_only = kwargs.get("numeric_only", False)
+        if not numeric_only:
+            for col, dt in self._query_compiler.dtypes.items():
+                if is_datetime64_any_dtype(dt):
+                    if self._df[col].hasnans:
+                        fallback = True
+                        break
+                    else:
+                        converted_columns[col] = dt
+        if fallback:
+            # We cannot use map-reduce approach because non-numeric columns are present and
+            # are requested to be processed. It happens because in map-reduce we use "sum"
+            # aggregation which always drops non-numeric columns unconditionally, no matter
+            # what arguments are specified. It is different from how "mean" handles non-numeric
+            # columns (in particular datetime types). We could use a workaround to convert
+            # datetime to int64 but it works only when NaNs aren't present in datetime columns.
+            # NaNs converted to int64 produce wrong results and have to be handled differently,
+            # so we have to resort to less efficient approach of broadcast full axis in
+            # _apply_agg_function.
+            result = self._apply_agg_function(lambda df: df.mean(*args, **kwargs))
+        else:
+            if len(converted_columns) > 0:
+                # Convert all datetime64 types to int64 to allow pandas "sum" to work.
+                self._df = self._df.astype(
+                    {col: "int64" for col in converted_columns.keys()}
+                )
```

Comment:
I still think that it's pandas-backend specific because of its MapReduce implementation and so should be moved to the QC. Let's wait for other reviewers' opinions though.

Reply:
in pandas this astype would make a copy, so we'd do `.view('i8')` instead.  is that different in modin?

Reply:
> in pandas this astype would make a copy, so we'd do .view('i8') instead. is that different in modin?

No, `astype` creates a copy in Modin, too. Unfortunately, in Modin, `Series.view` does not have the same semantics as in pandas. While it creates a dataframe with the correct type, the new series doesn't share the same underlying values as the original one. Not only that, Modin `view` also creates a copy of all the data under the hood. I've created #4650 to track the differences in behavior.